### ISSUE TITLE
Failure with bdist_wheel and namespace packages

### ISSUE THREAD

**Issue by gdude2002** [Nuitka/Nuitka#470]
I'm relatively inexperienced with setuptools, but I'm attempting to create a module that makes use of new-style [native namespace packages](https://packaging.python.org/guides/packaging-namespace-packages/#native-namespace-packages). Most things work fine, including `sdist` - but `bdist_nuitka` fails with the following :

```
(venv) [gdude@ns348665 Sidecar]$ python setup.py bdist_nuitka
running bdist_nuitka
running build
running build_py
creating build
creating build/lib
creating build/lib/comect
creating build/lib/comect/sidecar
copying src/comect/sidecar/__init__.py -> build/lib/comect/sidecar
Traceback (most recent call last):
  File "setup.py", line 44, in <module>
    install_requires=["pygtail==0.11.0 "],
  File "/home/gdude/Comect/Sidecar/venv/lib/python3.7/site-packages/setuptools-40.8.0-py3.7.egg/setuptools/__init__.py", line 145, in setup
  File "/usr/lib/python3.7/distutils/core.py", line 148, in setup
    dist.run_commands()
  File "/usr/lib/python3.7/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/usr/lib/python3.7/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/home/gdude/Comect/Sidecar/venv/lib/python3.7/site-packages/wheel/bdist_wheel.py", line 192, in run
    self.run_command('build')
  File "/usr/lib/python3.7/distutils/cmd.py", line 313, in run_command
    self.distribution.run_command(command)
  File "/usr/lib/python3.7/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/home/gdude/Comect/Sidecar/venv/lib/python3.7/site-packages/nuitka/distutils/bdist_nuitka.py", line 60, in run
    self._buildPackage(os.path.abspath(self.build_lib))
  File "/home/gdude/Comect/Sidecar/venv/lib/python3.7/site-packages/nuitka/distutils/bdist_nuitka.py", line 85, in _buildPackage
    assert package is None, package
AssertionError: comect
```

Running with `-O` will skip asserts, so for the hell of it I gave that a try too.

```
(venv) [gdude@ns348665 Sidecar]$ python -O setup.py bdist_nuitka
running bdist_nuitka
running build
running build_py
Nuitka:INFO:Recurse to import 'comect.sidecar' from 'comect/sidecar'. (Module is exact match of 'comect.sidecar' instructed by user to recurse to.)
Nuitka:INFO:Recurse to import 'comect' from 'comect'. (Containing package of recursed module 'comect.sidecar'.)
Nuitka:INFO:Included compiled module 'comect'.
Nuitka:INFO:Included compiled module 'comect.sidecar'.
Nuitka:INFO:Included compiled module 'sidecar'.
error: [Errno 2] No such file or directory: '/home/gdude/Comect/Sidecar/build/lib/comect.sidecar'
```

This results in the following file structure:

![Screenshot of build/ directory](https://user-images.githubusercontent.com/204153/61794285-c8a1b600-ae18-11e9-8706-f70e4fbc837b.png)

It looks to me like a simple enough issue, but I'm not 100% sure whether I'm doing something wrong. Any ideas?

---

**Environment details**

Python: **3.7.3**
Nuitka: **0.6.4** (Installed with pip into a virtualenv)
Platform: **Arch Linux**, x86_64

---

Example code can be found [at this commit](https://github.com/comect/Sidecar/tree/d4ce7e7c4da0191d0f318ecb6ebf9449ca84dbc2) - this is the actual project I'm working with.

Thanks in advance!

**Comment by kayhayen** [comment]
Checkout this here:

https://github.com/Nuitka/Nuitka/blob/4f6db834d39037a9f7241fe5e9cd7e6c0aca76a9/nuitka/distutils/bdist_nuitka.py#L75

Maybe if you figure out package and module_name properly, that is already about it. Can you give it a try?

**Comment by gdude2002** [comment]
I'm taking a look now, but as I say, I'm not familiar with the internals of setuptools/disutils at all, so I'm not entirely sure how much help I'm gonna be here.

Here's the list containing the generated command to call Nuitka, used here...

https://github.com/Nuitka/Nuitka/blob/4f6db834d39037a9f7241fe5e9cd7e6c0aca76a9/nuitka/distutils/bdist_nuitka.py#L141

![Screenshot of debugger](https://user-images.githubusercontent.com/204153/61829994-07a82980-ae62-11e9-8e9d-14dba921b001.png)

Stepping over this line, we get this:

```
No module named /home/gdude/PycharmProjects/Comect/Sidecar/venv/lib/python3.7/site-packages/nuitka/__main__.py
```

I honestly don't understand where this error is coming from specifically (and it's a subprocess so I'm not sure I could attach a debugger anyway) - but I can verify that the file exists for sure.

```bash
[gdude@manjaro ~]$ file /home/gdude/PycharmProjects/Comect/Sidecar/venv/lib/python3.7/site-packages/nuitka/__main__.py
/home/gdude/PycharmProjects/Comect/Sidecar/venv/lib/python3.7/site-packages/nuitka/__main__.py: Python script, ASCII text executable
```

I added `--verbose` to the list using my debugger, but the output was the same. Any idea what that suggests?

So, what happens if we run the generated command ourselves? Well..

```
(venv) [gdude@manjaro lib]$ python -m nuitka --module --plugin-enable=pylint-warnings --output-dir=/home/gdude/PycharmProjects/Comect/Sidecar/build/lib --include-package=comect.sidecar --nofollow-import-to=*.tests --show-modules --remove-output /home/gdude/PycharmProjects/Comect/Sidecar/build/lib/comect/sidecar
Nuitka:INFO:Recurse to import 'comect.sidecar' from 'comect/sidecar'. (Module is exact match of 'comect.sidecar' instructed by user to recurse to.)
Nuitka:INFO:Recurse to import 'comect' from 'comect'. (Containing package of recursed module 'comect.sidecar'.)
Nuitka:INFO:Included compiled module 'comect'.
Nuitka:INFO:Included compiled module 'comect.sidecar'.
Nuitka:INFO:Included compiled module 'sidecar'.
```

Curiouser and curiouser. As for the `No such file or directory` exception - that's raised here:

https://github.com/Nuitka/Nuitka/blob/4f6db834d39037a9f7241fe5e9cd7e6c0aca76a9/nuitka/distutils/bdist_nuitka.py#L152

Indeed, inspecting the arguments, we've got:

```py
{
    "self.build_lib": '/home/gdude/PycharmProjects/Comect/Sidecar/build/lib'
    "self.main_package": 'comect.sidecar'
}
```

This results, of course, in `/home/gdude/PycharmProjects/Comect/Sidecar/build/lib/comect.sidecar` being removed, which doesn't exist.

Let's try updating this line as a stop-gap.

```py
shutil.rmtree(os.path.join(self.build_lib, self.main_package.replace(".", "/")))
```

Here's what we get:

![Screenshot of output](https://user-images.githubusercontent.com/204153/61830926-58208680-ae64-11e9-8a79-ea875fd85882.png)

![Screenshot of wheel](https://user-images.githubusercontent.com/204153/61830949-666ea280-ae64-11e9-9b42-b2972c5ebb6d.png)

Success!

---

So, my thought process now is that the actual exception itself is a pretty easy solve - we just need to make sure the path we delete is actually a path converted from possibly-dotted package names.

The assert, which you may have noticed I haven't addressed yet, I'm not too sure about. It appears that a built-in module here would result in `package` not being set to `None`, but clearly this isn't a safe assertion to make when it comes to namespace packages. This is better checked out by someone familiar with the Nuitka internals, though.

---

I hope that helps you guys out a bit.

**Comment by kayhayen** [comment]
The assertion was merely informative, that the code is not ready to handle packages in packages, but can be removed, if you implemented that.

I think the compilation has to give a "sidecar" module, a "sidecar.so" indeed, but within a "comect" namepace package.

**Comment by kayhayen** [comment]
Maybe can you make a PR, where you provide a minimal example like this in `tests/distutils` in the same way as the other cases? Then I can compare it with what CPython does, and see what needs to be changed. Now I doner what is the normal result @gdude2002

**Comment by gdude2002** [comment]
Sure thing, I'll see about getting that PR'd today, assuming things work out.

**Comment by kayhayen** [comment]
This is solved on develop branch.

**Comment by gdude2002** [comment]
Thanks @kayhayen, I'll give it a test when I get home tonight!

Since I'm sure it passes the test though, I'd be fine if this ticket was closed before that.

**Comment by kayhayen** [comment]
We close issues when it's in a stable release, this is "Done" in planning though, this way people with the same issue, can tell that it's not in there yet.

**Comment by kayhayen** [comment]
Released as part of 0.6.6
